File: ProblemSet1 - C /cash/cash.c
#include <cs50.h>
#include <stdio.h>

int get_cents(void);
int calculate_quarters(int cents);
int calculate_dimes(int cents);
int calculate_nickels(int cents);
int calculate_pennies(int cents);

int main(void)
{
    // Prompt the user for change owed, in cents
    int cents = get_cents();

    // Calculate how many quarters you should give customer
    int quarters = calculate_quarters(cents);
    // Subtract the value of those quarters from cents
    cents = cents - quarters * 25;

    // Calculate how many dimes you should give customer
    int dimes = calculate_dimes(cents);
    // Subtract the value of those dimes from remaining cents
    cents = cents - dimes * 10;

    // Calculate how many nickels you should give customer
    int nickels = calculate_nickels(cents);
    // Subtract the value of those nickels from remaining cents
    cents = cents - nickels * 5;

    // Calculate how many pennies you should give customer
    int pennies = calculate_pennies(cents);
    // Subtract the value of those pennies from remaining cents
    cents = cents - pennies * 1;

    // Sum the number of quarters, dimes, nickels, and pennies used
    int coins = quarters + dimes + nickels + pennies;

    // Print that sum
    printf("%i\n", coins);
}
// Function to prompt the user for a non-negative amount of change in cents
int get_cents(void)
{
    int cents;
    do
    {
        cents = get_int("Change owed: ");
    }
    while (cents < 0);
    return cents;
}
// Function to calculate the number of quarters that can be given for the specified cents
int calculate_quarters(int cents)
{
    int quarters = 0;
    while (cents >= 25)
    {
        cents = cents - 25;
        quarters++;
    }
    return quarters;
}
// Function to calculate the number of dimes that can be given for the specified cents
int calculate_dimes(int cents)
{
    int dimes = 0;
    while (cents >= 10)
    {
        cents = cents - 10;
        dimes++;
    }
    return dimes;
}
// Function to calculate the number of nickles that can be given for the specified cents
int calculate_nickels(int cents)
{
    int nickels = 0;
    while (cents >= 5)
    {
        cents = cents - 5;
        nickels++;
    }
    return nickels;
}
// Function to calculate the number of pennies that can be given for the specified cents
int calculate_pennies(int cents)
{
    int pennies = 0;
    while (cents >= 1)
    {
        cents = cents - 1;
        pennies++;
    }
    return pennies;
}

File: ProblemSet6 - Python/sentimental-cash/cash.py
from cs50 import get_float

def main():
    change = get_change_owed("Change owed: ")
    cents = dollars_to_cents(change)

    # Prompt the user for change owed, in cents
    print(cents_to_coins(cents))

# Prompt user for an amount of change
def get_change_owed(prompt):
    # Start an infinite loop to repeatedly ask for input until a valid value is received
    while True:
        # Call the function get_float with the provided prompt to get user input
        n = get_float(prompt)
        # Check if the returned value n is greater than 0
        if n > 0:
            # If n is valid (greater than 0), exit the loop
            break
    # Return the valid positive float value n
    return n

# Convert dollars to cents
def dollars_to_cents(amount):
    c = round(amount * 100)
    return c

# Convert an amount in cents into the number of coins (quarters, dimes, nickels, and pennies).
def cents_to_coins(cents):
    # Calculate how many quarters are in the total cents.
    quarters = cents // 25
    # Calculate how many dimes are left after removing quarters.
    dimes = (cents % 25) // 10
    # Calculate how many nickels are left after removing quarters and dimes.
    nickels = ((cents % 25) % 10) // 5
    # Calculate how many pennies are left after removing quarters, dimes, and nickels.
    pennies = ((cents % 25) % 10) % 5
    # Return the total number of coins as a sum of all types.
    return quarters + dimes + nickels + pennies

if __name__ == "__main__":
    main()
